persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/MutationMyBatisRepository.java

```diff
@@ -43,8 +44,36 @@ public MutationMeta getMetaMutationsInMolecularProfileBySampleListId(String mole
                                                                   String projection, Integer pageSize, 
                                                                   Integer pageNumber, String sortBy, String direction) {
 
-        return mutationMapper.getMutationsInMultipleMolecularProfiles(molecularProfileIds, sampleIds, entrezGeneIds, 
-            null, projection, pageSize, offsetCalculator.calculate(pageSize, pageNumber), sortBy, direction);
+        return getGroupedMolecularProfileSamples(molecularProfileIds, sampleIds)
+            .entrySet()
+            .stream()
+            .flatMap(entry -> fetchMutationsInMolecularProfile(entry.getKey(),
+                entry.getValue(),
+                entrezGeneIds,
+                null,
+                projection,
+                pageSize,
+                offsetCalculator.calculate(pageSize, pageNumber),
+                sortBy,
+                direction).stream())
+            .collect(Collectors.toList());
+    }
+
+    private Map<String,List<String>> getGroupedMolecularProfileSamples(List<String> molecularProfileIds, List<String> sampleIds) {
+        Map<String,List<String>> groupMolecularProfileSamples = new HashMap<>();
+
+        for(int i = 0; i< molecularProfileIds.size(); i++) {
+            String molecularProfileId = molecularProfileIds.get(i);
+            String sampleId = sampleIds.get(i);
+            if(!groupMolecularProfileSamples.containsKey(molecularProfileId)) {
+                List<String> sampleList = new ArrayList<>();
```

Comment:
Just a thought : if a query is made where the same (identical sample, identical profile) sample is listed multiple times, then what we have decided to do is to return a response which shows multiple occurrences of that sample? (otherwise we could use a hashset instead of arraylist)

And second thought : by using a map of molecularprofile groups, is the ordering of response records altered? Was the order arbitrary before or is maintaining the order of response records important?

Reply:
> Just a thought : if a query is made where the same (identical sample, identical profile) sample is listed multiple times, then what we have decided to do is to return a response which shows multiple occurrences of that sample? (otherwise we could use a hashset instead of arraylist)

We just need to return distinct response(mysql do that uniqueness). Yes, we could replace it by `Set`
> 
> And second thought : by using a map of molecularprofile groups, is the ordering of response records altered? Was the order arbitrary before or is maintaining the order of response records important?

Order is not important

Reply:
Ok, updated it to `Set`